torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -1075,14 +1025,58 @@ void TensorPipeAgent::markFutureWithError(
                      futureMessage{std::move(futureMessage)},
                      errorMsg{std::move(errorMsg)}]() mutable {
       futureMessage->futMsg.setError(std::move(errorMsg));
-      // The future's callbacks may schedule further RPCs, increasing the count.
-      // Thus we must decrease it after completing the future, otherwise it may
-      // briefly dip to zero and trick join into thinking all work is done.
-      decreaseCallCount(clientActiveCalls_);
```

Comment:
It seems that `markFutureWithError` is also called when a timeout expires, which means we used to decrease the count when that happened, but now we don't do it anymore. Was that intended behavior? If so, we should try to preserve it...